archive/100-theorems-list/37_solution_of_cubic.lean

```diff
@@ -0,0 +1,174 @@
+          t * s = p ∧
+          (p = 0 → t = 0))
+  (x : ℂ) :
+   x^3 + 3 * p * x - 2 * q = 0 ↔
+      x = s       - t        ∨
+      x = s * ω   - t * ω^2  ∨
+      x = s * ω^2 - t * ω    :=
+begin
+  have h₁ : ∀ x a₁ a₂ a₃ : ℂ, x = a₁ ∨ x = a₂ ∨ x = a₃ ↔ (x - a₁) * (x - a₂) * (x - a₃) = 0,
+  { intros,
+    iterate 2 { rw [mul_eq_zero] },
+    iterate 3 { rw [sub_eq_zero] },
+    rw or.assoc, },
+  rw h₁,
+  suffices : x ^ 3 + 3 * p * x - 2 * q
+    = (x - (s - t)) * (x - (s * ω - t * ω^2)) * (x - (s * ω^2 - t * ω)),
+    { rw this, },
+  obtain ⟨hr, hs3, ht, hsp⟩ := h_eq,
+  have hc : s^3 - t^3 = 2 * q,
+    { by_cases hp : p = 0,
+
+      { simp [hp] at hs3, rw [hsp hp, hs3], norm_num },
+      { simp [hp] at hs3, have : s ≠ 0 := λ h, by { apply hp, rw [h, mul_zero] at ht, exact ht.symm },
+        have h_nonzero: q + r ≠ 0 := by { rw ← hs3, exact pow_ne_zero _ this },
+        have hp3 : p^3 = r^2 - q^2 := by rw [hr]; ring,
+        calc    s^3 - t^3
+              = s^3 - p^3/s^3 : by { rw [← ht], field_simp, ring }
+          ... = (q+r) - (r^2-q^2)/(q+r) : by rw [hs3, hp3]
+          ... = (q+r) + (q-r) * ((q+r) / (q+r)): by ring
+          ... = 2 * q : by rw div_self h_nonzero; ring } },
+  symmetry,
+  calc (x - (s - t)) * (x - (s * ω - t * ω^2)) * (x - (s * ω^2 - t * ω))
+      = x^3 - (s-t) * (1+ω+ω^2) * x^2
+        + ((s^2+t^2)*ω*(1+ω+ω^2) - s*t*(-3 + 3*(1+ω+ω^2) + ω*(ω^3-1))) * x
+        - (s^3-t^3)*ω^3 + s*t*(s-t)*ω^2*(1+ω+ω^2)
+        : by ring
+  ... = x^3 + 3*(t*s)*x - (s^3-t^3)
+        : by { rw [omega_sum, omega_cube_eq_one], ring }
+  ... = x^3 + 3*p*x - 2*q : by rw [ht, hc]
+end
+
+/-- The solution of the monic cubic equation. -/
+lemma cubic_monic_eq_zero_iff
+  (h_eq : p = (3 * c - b^2) / 9 ∧
+          q = (9 * b * c - 2 * b^3 - 27 * d) / 54 ∧
+          r^2 = q^2 + p^3 ∧
+          s^3 = (if p = 0 then 2 * q else q + r) ∧
+          t * s = p ∧
+          (p = 0 → t = 0))
+  (x : ℂ) :
+      x^3 + b * x^2 + c * x + d = 0 ↔
+        x = s       - t       - b / 3 ∨
+        x = s * ω   - t * ω^2 - b / 3 ∨
+        x = s * ω^2 - t * ω   - b / 3 :=
+begin
+  let y := x + b / 3,
+  have h₁ : x^3 + b * x^2 + c * x + d = y^3 + 3 * p * y - 2 * q,
+  { dsimp [y],
+    rcases h_eq with ⟨hp, hq, -⟩,
+    rw [hp, hq],
+    field_simp, ring, },
+  have h₂ := @cubic_basic_eq_zero_iff p q r s t _ (by simp [h_eq] at h_eq ⊢; cc) y,
+  rw [h₁, h₂],
+  dsimp [y],
+  have h₃ : ∀ s t : ℂ, x + b / 3 = s - t ↔ x = s - t - b / 3,
+    { intros s t, split; intros h,
+      { calc x = x + b / 3 - b / 3 : by ring
+         ... = s - t - b / 3 : by rw h },
+      { rw h, ring } },
+  repeat { rw h₃ },
+end
+
+/-- **The Solution of Cubic** -/
+theorem cubic_eq_zero_iff (ha : a ≠ 0)
+  (h_eq : p = (3 * a * c - b^2) / (9 * a^2) ∧
+          q = (9 * a * b * c - 2 * b^3 - 27 * a^2 * d) / (54 * a^3) ∧
+          r^2 = q^2 + p^3 ∧
+          s^3 = (if p = 0 then 2 * q else q + r) ∧
+          t * s = p ∧
+          (p = 0 → t = 0))
```

Comment:
You should split this hypothesis into six, one for each proposition that you take the `and` of. Also maybe you could have one theorem for `p = 0` and one for `p ≠ 0`?

Reply:
Is it good idea to join variable and corresponding hypothesis `(hr : ∃ r : ℂ, r^2 = q^2 + p^3)` with existential quantifier?  
Currently I'm using `{p q r ... : ℂ}, ... (hr :  r^2 = q^2 + p^3)`